examples/99_in_tool_selection/ex_in_tool_selection/main.rb

```diff
@@ -0,0 +1,364 @@
+      end
+
+      # Check if the selection as a whole is valid for this tool.
+      def valid_selection?
+        selection = Sketchup.active_model.selection
+
+        selection.size == 1 && valid_entity?(selection.first)
+      end
+
+      # Try selecting the entity under the mouse cursor, if the tool allows it.
+      def try_select_entity(view, x, y)
+        view.model.selection.clear
+
+        pickhelper = view.pick_helper(x, y)
+        @hovered_entity = pickhelper.best_picked
+        return unless @hovered_entity
+
+        @valid_hovered_entity = valid_entity?(@hovered_entity)
+        return unless @valid_hovered_entity
+
+        view.model.selection.add(@hovered_entity)
+      end
+
+      def start_move(view)
+        # HACK: create a new input point with a given position rather than
+        # copying the existing one. This fixes an issue in some cases where
+        # the points appear to get linked and the @picked_first_ip position
+        # changes when you move the mouse around, even if that should only
+        # affect @mouse_ip.
+        ### @picked_first_ip.copy!(@mouse_ip)
+        @picked_first_ip = Sketchup::InputPoint.new(@mouse_ip.position)
+
+        # Wrap model changes in an operation so it can be undone in one step.
+        # Typically we disable drawing in operations to increase performance,
+        # but here we want live updates when moving the mouse.
+        view.model.start_operation("Move", false)
+
+        # Track whether we are in an operation so we can abort it if the user
+        # activates another tool.
+        @in_operation = true
+
+        # Track the mouse position between frames so we can update the position
+        # of the selected entities.
+        @last_frame_mouse_position = @mouse_ip.position
+      end
+
+      def do_move(view)
+        return unless @last_frame_mouse_position
+
+        # Move entities the same distance input point moved since last frame.
+        frame_move = @mouse_ip.position - @last_frame_mouse_position
+        @last_frame_mouse_position = @mouse_ip.position
+
+        view.model.active_entities.transform_entities(frame_move, @selection)
+      end
+
+      def end_move(view)
+        view.model.commit_operation
+        @in_operation = false
+
+        @picked_first_ip.clear
+        @inference_lock_helper.unlock
+
+        # If tool was activated with a pre-selection, keep the selection and let
+        # the user pick a starting point anywhere. Otherwise, drop the selection
+        # and select whatever the user clicks when starting next move operation.
+        reset_tool unless @preselection
+      end
+
+      def pick_mouse_position(view)
+        if !picked_first_point?
+          @mouse_ip.pick(view, @mouse_position.x, @mouse_position.y)
+        else
+          # If we could, we would want the InputPoint to ignore entities in
+          # @selection. We are currently getting some quite strange behavior
```

Comment:
"quite strange behavior" could do with some explanation here. People read this to learn and will puzzle to what this means. (I do too.)

Reply:
Clarified the comment.